### PR TITLE
AzureMonitor: Add support for selecting multiple options when using the equals and not equals dimension filters

### LINKED ISSUE #48553: AzureMonitor: Add support for selecting multiple Dimension options when using the 'eq' and 'ne' operators
The Azure portal currently allows users to select multiple options when applying a filter that makes use of the equals and not equals operators.

### PR DESCRIPTION
**Pull request by aangelisc**
**What this PR does / why we need it**:

Adds support for selecting multiple dimension values when using the equals and not equals operators in the Azure Monitor datasource.

**Which issue(s) this PR fixes**:

Fixes #48553 

Also fixes a minor bug where the labels would be cleared on selection.

### CODE REVIEW

File: pkg/tsdb/azuremonitor/metrics/azuremonitor-datasource.go [hunk 1]
```diff
@@ -129,10 +129,12 @@ func (e *AzureMonitorDatasource) buildQueries(queries []backend.DataQuery, dsInf
 			dimSB.WriteString(fmt.Sprintf("%s eq '%s'", dimension, dimensionFilter))
 		} else {
 			for i, filter := range azJSONModel.DimensionFilters {
-				if filter.Operator != "eq" && filter.Filter == "*" {
+				if filter.Operator != "eq" && filter.Filter == "*" && len(filter.Filters) == 0 {
```
Comment (iwysiu) [inline pkg/tsdb/azuremonitor/metrics/azuremonitor-datasource.go:132]: `Filters` doesn't get cleared when it switches to `starts with`, so it gets a `400 Bad Request` if you:
1) Set a `eq` operation with 2 values
2) switch to the `starts with` operation

**Review by iwysiu — COMMENTED** [review]
I don't think it's necessarily wrong, but it seems weird to have `eq` and `ne` using the same field and `sw` using a different one. Making separate filters for each might make more sense.

File: pkg/tsdb/azuremonitor/metrics/azuremonitor-datasource.go (hunk 1 above)
Comment (aangelisc) [reply pkg/tsdb/azuremonitor/metrics/azuremonitor-datasource.go:132]: Great catch here, have fixed this 😊

**Comment by aangelisc** [comment]
> I don't think it's necessarily wrong, but it seems weird to have `eq` and `ne` using the same field and `sw` using a different one. Making separate filters for each might make more sense.

I think it's acceptable in this case as the logic for the select is the same across both, there's no real difference between the fields for either operator. I'm also weary of duplicating code if it can be succinct 🤔

File: pkg/tsdb/azuremonitor/types/types.go [hunk 2]
```diff
@@ -139,9 +140,10 @@ type AzureMonitorJSONQuery struct {
 // AzureMonitorDimensionFilter is the model for the frontend sent for azureMonitor metric
 // queries like "BlobType", "eq", "*"
 type AzureMonitorDimensionFilter struct {
-	Dimension string `json:"dimension"`
-	Operator  string `json:"operator"`
-	Filter    string `json:"filter"`
+	Dimension string   `json:"dimension"`
+	Operator  string   `json:"operator"`
+	Filter    string   `json:"filter"`
+	Filters   []string `json:"filters,omitempty"`
```
Comment (andresmgot) [inline pkg/tsdb/azuremonitor/types/types.go:146]: it doesn't sound right to me that we have Filter and Filters. Is that necessary? Why is not a Filter a slice of one element? It sounds like you are doing that to support the current queries but if that's the case we would need to:
 - Add a comment saying that using `filter` is deprecated, and migrate here to use `filters` (wherever this struct is being populated).
 - In the QueryEditor, migrate it too so after editing a query, it is now using the list.

Does that make sense? That way we can remove the special handling when Filter is set.

Comment (aangelisc) [reply pkg/tsdb/azuremonitor/types/types.go:146]: A migration would make the most sense here to support current queries and also implement the new functionality. I'll look into this. This would simplify some things as the `Filter` property is essentially only used now for the `sw` operator.

Comment (andresmgot) [reply pkg/tsdb/azuremonitor/types/types.go:146]: cool, let me know if you find any issue with that.

File: pkg/tsdb/azuremonitor/types/types.go [hunk 3]
```diff
@@ -139,17 +140,27 @@ type AzureMonitorJSONQuery struct {
...
 }
 
-func (a AzureMonitorDimensionFilter) String() string {
-	filter := "*"
-	if a.Filter != "" {
-		filter = a.Filter
+func (a AzureMonitorDimensionFilter) ConstructFiltersString() string {
+	if len(a.Filters) == 1 {
+		return fmt.Sprintf("%v %v '%v'", a.Dimension, a.Operator, a.Filters[0])
+	} else {
```
Comment (andresmgot) [inline pkg/tsdb/azuremonitor/types/types.go:153]: I don't think you need this `if` (the result would be the same with the `else` logic as far as I can see)

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts [hunk 4]
```diff
@@ -106,10 +106,12 @@ export default class AzureMonitorDatasource extends DataSourceWithBackend<AzureM
       .filter((f) => f.dimension && f.dimension !== 'None')
       .map((f) => {
         const filter = templateSrv.replace(f.filter ?? '', scopedVars);
+        const filters = f.filters?.map((filter) => templateSrv.replace(filter ?? '', scopedVars));
         return {
           dimension: templateSrv.replace(f.dimension, scopedVars),
           operator: f.operator || 'eq',
-          filter: filter || '*', // send * when empty
+          filter: filter || '*', // send * when empty,
```
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts:113]: I don't think you need to setup a default `filter` since it's now deprecated, the behavior should be moved to the `filters`

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx [hunk 5]
```diff
@@ -44,7 +44,11 @@ const useDimensionLabels = (data: PanelData | undefined, query: AzureMonitorQuer
     }
     setDimensionLabels((prevLabels) => {
       const newLabels: DimensionLabels = {};
-      for (const label of Object.keys(labelsObj)) {
+      const currentLabels = Object.keys(labelsObj);
+      if (currentLabels.length === 0) {
+        return prevLabels;
+      }
```
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:50]: this change seems unrelated to the PR?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx [hunk 6]
```diff
@@ -146,16 +170,27 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ data, query, dimensio
...
+                menuShouldPortal
+                placeholder="Select value(s)"
+                value={filter.filters}
+                options={getValidMultiSelectOptions(filter.filters, filter.dimension)}
+                onChange={(v) => onMultiSelectFilterChange(index, v)}
+                aria-label={'dimension-labels-select'}
+                allowCustomValue
+              />
+            ) : (
+              <Select
```
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:184]: I am not sure why you can only set one filter when using  `starts with`? Is that an API limitation? (Maybe you can explain that in a comment)

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.ts [hunk 7]
```diff
@@ -22,6 +22,7 @@ export default function migrateQuery(query: AzureMonitorQuery): AzureMonitorQuer
   workingQuery = migrateToDefaultNamespace(workingQuery);
   workingQuery = migrateMetricsDimensionFilters(workingQuery);
   workingQuery = migrateResourceUri(workingQuery);
+  workingQuery = migrateDimensionFilters(workingQuery);
```
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.ts:25]: there is already a `migrateMetricsDimensionFilters` above. Can we reuse that method or is there any reason for having a different function here?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.test.ts [hunk 8]
```diff
@@ -115,4 +115,103 @@ describe('AzureMonitor: migrateQuery', () => {
...
+              },
+            ],
+          }),
+        })
+      );
+    });
+    it('correctly merges values', () => {
+      const dimensionFilters: AzureMetricDimension[] = [
+        { dimension: 'TestDimension', operator: 'eq', filter: 'testFilter2', filters: ['testFilter'] },
+      ];
```
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.test.ts:200]: I think that the behavior should be to ignore `filter` if `filters` is set. There should not be a case in which both are used, queries should be either migrated / new or non-migrated.

**Review by andresmgot — COMMENTED** [review]
Thanks for the changes @aangelisc ! I have some more suggestions, let me know what you think.

File: pkg/tsdb/azuremonitor/types/types.go (hunk 3 above)
Comment (aangelisc) [reply pkg/tsdb/azuremonitor/types/types.go:153]: Yep will update this 👍

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx (hunk 5 above)
Comment (aangelisc) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:50]: This was a bug fix for an issue I noticed with the labels. As it was so short I figured I'd just add it here.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx (hunk 6 above)
Comment (aangelisc) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:184]: This is an API limitation, will add a comment to explain.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.ts (hunk 7 above)
Comment (aangelisc) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.ts:25]: I personally think the migrations should be kept independent of each other as it makes it easier to track what changes are being made to the query and when. Happy to discuss this and hear opinions on it 😊

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts [hunk 9]
```diff
@@ -105,12 +105,10 @@ export default class AzureMonitorDatasource extends DataSourceWithBackend<AzureM
     const dimensionFilters = (item.dimensionFilters ?? [])
       .filter((f) => f.dimension && f.dimension !== 'None')
       .map((f) => {
-        const filter = templateSrv.replace(f.filter ?? '', scopedVars);
```
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts:108]: can you double check if you can remove this filter? my question is because I don't know if queries that reach this function has been migrated already or not. I believe they will but better to be sure.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx [hunk 10]
```diff
@@ -181,6 +181,7 @@ const DimensionFields: React.FC<DimensionFieldsProps> = ({ data, query, dimensio
                 allowCustomValue
               />
             ) : (
+              // The API does not currently allow for multiple starts with clauses to be used.
```
Comment (andresmgot) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:184]: nit:
```suggestion
              // The API does not currently allow for multiple "starts with" clauses to be used.
```

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.ts (hunk 7 above)
Comment (andresmgot) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.ts:25]: I agree, better to have different methods but here it would be helpful to know what's the difference between them. Maybe renaming would help here: the one in line 23 can be  `migrateDimensionToDimensionFilter` and this one `migrateDimensionFilterToArray`. WDYT?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx (hunk 5 above)
Comment (andresmgot) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/DimensionFields.tsx:50]: that's fine for small patches :+1: What I usually do with these is either add a comment to my own PR review to highlight it or mention it in the PR description so the reviewer knows (and it's clear is not some leftover from the development)

**Review by andresmgot — APPROVED** [review]
LGTM! Just make sure that legacy queries, with a template variable in the filter still work after this :trophy:

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts (hunk 9 above)
Comment (aangelisc) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/azure_monitor_datasource.ts:108]: A query that gets to this point will be migrated first based on hitting [this line](https://github.com/grafana/grafana/blob/7b9929fffe88d81dd0a5dd1397bbd85e88fb9fe0/public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts#L73) and then [this line](https://github.com/grafana/grafana/blob/7b9929fffe88d81dd0a5dd1397bbd85e88fb9fe0/packages/grafana-runtime/src/utils/DataSourceWithBackend.ts#L143).

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.ts (hunk 7 above)
Comment (aangelisc) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/migrateQuery.ts:25]: I think a rename is more appropriate in this case 😊
